### PR TITLE
Make `prefer-for-of` block scope aware

### LINKED ISSUE #1884: Adding prefer-for-of option breaks tslint
### Bug Report

- __TSLint version__: `4.1.0`
- __TypeScript version__: `2.1.4`
- __Running TSLint via__: gulp-tslint

#### TypeScript code being linted

very large application...

with `tslint.json` configuration:

```
"prefer-for-of": true
```

#### Actual behavior

Fails after adding `prefer-for-of`:

```
TypeError: Cannot read property 'onlyArrayReadAccess' of undefined
    at PreferForOfWalker.visitForStatement (C:\<project>\node_modules\tslint\lib\rules\preferForOfRule.js:73:33)
    at PreferForOfWalker.SyntaxWalker.visitNode (C:\<project>\node_modules\tslint\lib\language\walker\syntaxWalker.js:352:22)
    at C:\<project>\node_modules\tslint\lib\language\walker\syntaxWalker.js:517:63
    at visitEachNode (C:\<project>\node_modules\typescript\lib\typescript.js:13907:30)
    at Object.forEachChild (C:\<project>\node_modules\typescript\lib\typescript.js:14076:24)
    at PreferForOfWalker.SyntaxWalker.walkChildren (C:\<project>\node_modules\tslint\lib\language\walker\syntaxWalker.js:517:12)
    at PreferForOfWalker.SyntaxWalker.visitBlock (C:\<project>\node_modules\tslint\lib\language\walker\syntaxWalker.js:47:14)
    at PreferForOfWalker.SyntaxWalker.visitNode (C:\<project>\node_modules\tslint\lib\language\walker\syntaxWalker.js:289:22)
    at C:\<project>\node_modules\tslint\lib\language\walker\syntaxWalker.js:517:63
    at visitNode (C:\<project>\node_modules\typescript\lib\typescript.js:13895:20)

```

#### Expected behavior

Work as usual

### PR DESCRIPTION
**Pull request by nchen63**
fixes #1884

### CODE REVIEW

File: src/rules/preferForOfRule.ts [hunk 1]
```diff
@@ -46,24 +46,31 @@ interface IIncrementorState {
...
         super(sourceFile, options);
-        this.incrementorMap = {};
+    }
+
+    public createScope() {
+        return {};
+    }
+
+    public createBlockScope() {
+        // a map of incrementors and whether or not they are only used to index into an array reference in the for loop
```
Comment (adidahiya) [inline src/rules/preferForOfRule.ts]: this comment should go on the `IncrementorMap` definition

File: src/rules/preferForOfRule.ts [hunk 2]
```diff
@@ -46,24 +46,31 @@ interface IIncrementorState {
     onlyArrayReadAccess: boolean;
 }
 
-class PreferForOfWalker extends Lint.RuleWalker {
-    // a map of incrementors and whether or not they are only used to index into an array reference in the for loop
-    private incrementorMap: { [name: string]: IIncrementorState };
-
+type IncrementorMap = { [name: string]: IIncrementorState };
+class PreferForOfWalker extends Lint.BlockScopeAwareRuleWalker<{}, IncrementorMap> {
     constructor(sourceFile: ts.SourceFile, options: Lint.IOptions) {
```
Comment (adidahiya) [inline src/rules/preferForOfRule.ts]: you can omit this constructor now, it's not doing anything

File: src/rules/preferForOfRule.ts [hunk 3]
```diff
@@ -46,24 +46,31 @@ interface IIncrementorState {
     onlyArrayReadAccess: boolean;
 }
 
-class PreferForOfWalker extends Lint.RuleWalker {
-    // a map of incrementors and whether or not they are only used to index into an array reference in the for loop
-    private incrementorMap: { [name: string]: IIncrementorState };
-
+type IncrementorMap = { [name: string]: IIncrementorState };
+class PreferForOfWalker extends Lint.BlockScopeAwareRuleWalker<{}, IncrementorMap> {
```
Comment (adidahiya) [inline src/rules/preferForOfRule.ts]: nit: newline above this

File: src/language/walker/blockScopeAwareRuleWalker.ts [hunk 4]
```diff
@@ -58,6 +58,16 @@ export abstract class BlockScopeAwareRuleWalker<T, U> extends ScopeAwareRuleWalk
...
     }
 
+    public findBlockScope(predicate: (scope: U) => boolean) {
+        // look through block scopes from local -> global
+        for (let i = this.blockScopeStack.length - 1; i >= 0; i--) {
+            if (predicate(this.blockScopeStack[i])) {
+                return this.blockScopeStack[i];
+            }
+        }
+        return null;
```
Comment (adidahiya) [inline src/language/walker/blockScopeAwareRuleWalker.ts]: don't return null explicitly; return undefined instead

File: src/language/walker/blockScopeAwareRuleWalker.ts [hunk 5]
```diff
@@ -58,6 +58,16 @@ export abstract class BlockScopeAwareRuleWalker<T, U> extends ScopeAwareRuleWalk
         return;
     }
 
+    public findBlockScope(predicate: (scope: U) => boolean) {
```
Comment (adidahiya) [inline src/language/walker/blockScopeAwareRuleWalker.ts]: why do we need this? shouldn't it just be `this.getAllBlockScopes().find(predicate)`?

Comment (nchen63) [inline src/language/walker/blockScopeAwareRuleWalker.ts]: this goes in reverse order from local to global

**Comment by nchen63** [comment]
@adidahiya I'm pretty sure this fixes the issue. The only way the call stack could have occurred is if the same iterator name was reused (added twice, then deleted when exiting the "scope"). The choice seems to be to close the ticket with a likely fix (it can be reopened if it's not fixed), or leaving the ticket permanently open if the submitter doesn't update the ticket.

File: src/language/walker/blockScopeAwareRuleWalker.ts (hunk 5 above)
Comment (adidahiya) [inline src/language/walker/blockScopeAwareRuleWalker.ts]: Then you could call `reverse()` on the array; I think that'd be more readable

Comment (nchen63) [inline src/language/walker/blockScopeAwareRuleWalker.ts]: that creates a copy of the array. I think I made this change based on a discussion about not unnecessarily copying arrays.

Comment (adidahiya) [inline src/language/walker/blockScopeAwareRuleWalker.ts]: alright, that sounds reasonable. in the future, please leave a code comment

**Pull request merged**